### PR TITLE
Final experiments

### PR DESCRIPTION
**Pull request by dnachev**
- Implement a crude merge strategy. For simplicity and performance, the strategy will produce special changed keys format, where nested keys are represented as arrays, where the first entry is the name of the key in the parent object and the rest are the keys changed in the nested object.
  The `_didReceiveNestedProperties` will skip on detecting changes if the keys have been passed.
- Remove the `@` base model as it is no longer needed. This required adjusting the payloads a little, because I think they no longer require `projectionTypes` in meta (it doesn't bring any value). The most we may need there is a flag to not mark the model as loaded.
- Remove `pushInternalModel` hook which caused two tests to be skipped (`findRecord` and `peekRecord` one). The removal of the hook also required the switch to use `computeBaseModelName()` from the schema and `recordForId()`, which will always create a base record and it solves the problem of Ember Data creating an empty projection record for the purpose of fetching it, before we have received the base record.
- Skip tests, which rely on fetching updates to a projection. The serializer is supposed to always convert these to updates to the base model and the projection. This means the test for updates to base model are already covering the above case. Only thing, which these tests should catch are changes in the state of the projection record.
- Note, because of the merge strategy, other tests are failing. I'm going to look at them immediately to make sure they are no new bugs introduced there.

### CODE REVIEW

File: tests/unit/projection-test.js [hunk 1]
```diff
@@ -711,30 +718,21 @@ module('unit/projection', function(hooks) {
                 age: AUTHOR_AGE,
               },
             },
-            meta: {
-              projectionTypes: [BOOK_CLASS_PATH]
```
Comment (runspired) [inline tests/unit/projection-test.js]: This still has value because when it is present it alerts us that this data is partial data, we otherwise do not have a way of determining this. I was planning on using it's presence for developing the `load-data-but-leave-in-unloaded-state` behavior.

File: tests/unit/projection-test.js [hunk 2]
```diff
@@ -711,30 +718,21 @@ module('unit/projection', function(hooks) {
...
-            }
           },
         });
 
         projectedExcerpt = store.push({
           data: {
             id: BOOK_ID,
-            type: BOOK_CLASS_PATH,
```
Comment (runspired) [inline tests/unit/projection-test.js]: good catch

File: tests/unit/projection-test.js [hunk 3]
```diff
@@ -711,30 +718,21 @@ module('unit/projection', function(hooks) {
...
-            },
+            type: BOOK_EXCERPT_PROJECTION_CLASS_PATH,
             attributes: {},
           },
         });
 
         projectedPreview = store.push({
           data: {
             id: BOOK_ID,
-            type: BOOK_CLASS_PATH,
```
Comment (runspired) [inline tests/unit/projection-test.js]: another good catch :)

Comment (runspired) [inline tests/unit/projection-test.js]: another good catch :)

File: tests/unit/projection-test.js [hunk 4]
```diff
@@ -955,7 +953,9 @@ module('unit/projection', function(hooks) {
       assert.equal(get(projectedExcerpt, 'author.age'), AUTHOR_AGE, 'excerpt has the correct author.age');
     });
 
-    test('Updating an embedded object property on a projection updates the base-record and other projections', function(assert) {
+    // Skipped because we cannot really simulate an update to a projection, it is always an update to the base record
+    // only fetches are something we can distinguish
+    skip('Updating an embedded object property on a projection updates the base-record and other projections', function(assert) {
```
Comment (runspired) [inline tests/unit/projection-test.js]: I would unskip this, we can simulate an update to a projection if `projectionTypes` is our way of signaling `partial-data`. The payload should be update however to push the empty-projection as the primary record and the partial book in `included`

File: tests/unit/projection-test.js [hunk 5]
```diff
@@ -998,7 +998,7 @@ module('unit/projection', function(hooks) {
       assert.equal(get(projectedExcerpt, 'author.age'), NEW_AUTHOR_AGE, 'excerpt has the correct author.age');
     });
 
-    test('Updating an embedded object property on a nested projection updates the base-record and other projections', function(assert) {
+    skip('Updating an embedded object property on a nested projection updates the base-record and other projections', function(assert) {
```
Comment (runspired) [inline tests/unit/projection-test.js]: same as above RE unskip

File: tests/unit/projection-test.js [hunk 6]
```diff
@@ -1084,9 +1084,6 @@ module('unit/projection', function(hooks) {
             type: BOOK_CLASS_PATH,
             attributes: {
               publisher: `urn:${PUBLISHER_CLASS}:${PUBLISHER_ID}`,
-            },
-            meta: {
-              projectionTypes: [BOOK_CLASS_PATH]
```
Comment (runspired) [inline tests/unit/projection-test.js]: keep projectionTypes

File: tests/unit/projection-test.js [hunk 7]
```diff
@@ -1097,32 +1094,27 @@ module('unit/projection', function(hooks) {
                 name: PUBLISHER_NAME,
                 location: PUBLISHER_LOCATION,
                 owner: PUBLISHER_OWNER,
-              },
-              meta: {
-                projectionTypes: [PUBLISHER_CLASS, NORM_PROJECTED_PUBLISHER_CLASS]
```
Comment (runspired) [inline tests/unit/projection-test.js]: keep projectionTypes

File: tests/unit/projection-test.js [hunk 8]
```diff
@@ -1097,32 +1094,27 @@ module('unit/projection', function(hooks) {
...
               }
+            }, {
+              id: PUBLISHER_ID,
+              type: NORM_PROJECTED_PUBLISHER_CLASS,
+              attributes: {}
```
Comment (runspired) [inline tests/unit/projection-test.js]: The idea here is that the normalized payload would be roughly

```
{
  data: <TOP_LEVEL_PROJECTION>,
  included: [
    <PARTIAL_PRIMARY_RECORD>,
    <PARTIAL_RESOLVED_RECORD>,
    <RESOLVED_PROJECTION (empty)>
  ]
}
```

confirm?

File: tests/unit/projection-test.js [hunk 9]
```diff
@@ -1097,32 +1094,27 @@ module('unit/projection', function(hooks) {
...
           ]
         });
 
         projectedExcerpt = store.push({
           data: {
             id: BOOK_ID,
-            type: BOOK_CLASS_PATH,
-            meta: {
-              projectionTypes: [BOOK_EXCERPT_PROJECTION_CLASS_PATH],
-            },
```
Comment (runspired) [inline tests/unit/projection-test.js]: 👍

File: tests/unit/projection-test.js [hunk 10]
```diff
@@ -1097,32 +1094,27 @@ module('unit/projection', function(hooks) {
...
+            attributes: {}
           },
         });
 
         projectedPreview = store.push({
           data: {
             id: BOOK_ID,
-            type: BOOK_CLASS_PATH,
-            meta: {
-              projectionTypes: [BOOK_PREVIEW_PROJECTION_CLASS_PATH],
```
Comment (runspired) [inline tests/unit/projection-test.js]: 👍

File: tests/unit/projection-test.js [hunk 11]
```diff
@@ -1262,26 +1254,13 @@ module('unit/projection', function(hooks) {
...
-              type: PUBLISHER_CLASS,
-              attributes: {
-                location: NEW_PUBLISHER_LOCATION,
-                owner: NEW_PUBLISHER_OWNER
-              },
-              meta: {
-                projectionTypes: [PUBLISHER_CLASS]
-              }
-            }
-          ]
```
Comment (runspired) [inline tests/unit/projection-test.js]: semantically this is probably the same change as just dropping the erroneous inclusion of `projectionTypes`; however, since the idea is to simulate an included record in a different payload I'd prefer to keep it structured that way.

File: tests/unit/projection-test.js [hunk 12]
```diff
@@ -1307,6 +1286,7 @@ module('unit/projection', function(hooks) {
 
       run(() => {
         set(projectedExcerpt, 'publisher.location', NEW_PUBLISHER_LOCATION);
+        set(projectedExcerpt, 'publisher.owner', NEW_PUBLISHER_OWNER);
```
Comment (runspired) [inline tests/unit/projection-test.js]: I believe this should be an expect throws and the counts below should reflect that owner does not change since the publisher projection does not allow `owner`.

File: tests/unit/projection-test.js [hunk 13]
```diff
@@ -1352,7 +1332,7 @@ module('unit/projection', function(hooks) {
       assert.equal(get(projectedExcerpt, 'publisher.owner'), PUBLISHER_OWNER, 'excerpt has the correct publisher.owner');
     });
 
-    test('Updating a resolution property via a projection updates the base-record, other projections and nested projections', function(assert) {
+    skip('Updating a resolution property via a projection updates the base-record, other projections and nested projections', function(assert) {
```
Comment (runspired) [inline tests/unit/projection-test.js]: unskip as with above

File: tests/unit/projection-test.js [hunk 14]
```diff
@@ -1363,26 +1343,13 @@ module('unit/projection', function(hooks) {
       run(() => {
         store.push({
           data: {
-            id: BOOK_ID,
```
Comment (runspired) [inline tests/unit/projection-test.js]: I believe the correct change here is to move to the following format:

```
{
  data: <TOP_LEVEL_PROJECTION>,
  included: [
    <PARTIAL_PRIMARY_RECORD>,
    <PARTIAL_RESOLVED_RECORD>,
    <RESOLVED_PROJECTION (empty)>
  ]
}
```

File: tests/unit/projection-test.js [hunk 15]
```diff
@@ -1400,7 +1367,7 @@ module('unit/projection', function(hooks) {
       assert.equal(get(projectedExcerpt, 'publisher.owner'), NEW_PUBLISHER_OWNER, 'excerpt has the correct publisher.owner');
     });
 
-    test('Updating a resolution property via a nested projection updates the base-record, other projections', function(assert) {
+    skip('Updating a resolution property via a nested projection updates the base-record, other projections', function(assert) {
```
Comment (runspired) [inline tests/unit/projection-test.js]: also unskip

File: tests/unit/projection-test.js (hunk 1 above)
Comment (dnachev) [reply tests/unit/projection-test.js]: Should we just have a flag `partial: true`?

Comment (runspired) [reply tests/unit/projection-test.js]: Might be easier yes

File: tests/unit/projection-test.js (hunk 8 above)
Comment (dnachev) [reply tests/unit/projection-test.js]: Actually, you are correct. I will check the changes again to ensure I haven't just placed the base record as a top level, when it should be in included.

File: tests/unit/projection-test.js (hunk 11 above)
Comment (dnachev) [reply tests/unit/projection-test.js]: So, the idea is to always have book/projection of a book in the `data` and publisher in the `included`?

I was just trying to push the minimal required data to exercise the use case, in which case is an update to the publisher.

File: tests/unit/projection-test.js (hunk 12 above)
Comment (dnachev) [reply tests/unit/projection-test.js]: That's fair, didn't look at whether we have failing `set` for resolved entity.

File: tests/unit/projection-test.js [hunk 16]
```diff
@@ -619,22 +624,24 @@ module('unit/projection', function(hooks) {
     });
 
     test('Setting a projection updates the base-record and other projections', function(assert) {
-      let excerpt = this.records.projectedExcerpt;
+      let preview = this.records.projectedPreview;
```
Comment (runspired) [inline tests/unit/projection-test.js]: why this change?

File: tests/unit/projection-test.js (hunk 12 above)
Comment (runspired) [reply tests/unit/projection-test.js]: I mistook excerpt here for preview, this set should not be expected to throw

File: tests/unit/projection-test.js (hunk 16 above)
Comment (dnachev) [reply tests/unit/projection-test.js]: The excerpt doesn't whitelist the `chapter-1` property, which breaks. I figured `preview` suits the rest of the test conditions.
